src/base/http/connection.cpp

```diff
@@ -91,7 +92,22 @@ void Connection::read()
             return;
 
         case RequestParser::ParseStatus::OK: {
-                const Environment env {m_socket->localAddress(), m_socket->localPort(), m_socket->peerAddress(), m_socket->peerPort()};
+                QString forwarded_for = result.request.headers.value(Http::HEADER_X_FORWARDED_FOR);
```

Comment:
It's not valid place for this substitution. It should be done in some higher level (interested in this logic).

Reply:
WebApplication::isAuthNeeded() is better place.

Reply:
I was not sure where changes should be placed due to env variable being a const. I will try to move my code.

Reply:
>env variable being a const

Yes. It should be. You shouldn't change it. Just use your substitute value instead where it's applicable, i.e. in WebApplication::isAuthNeeded(), as I suggested before.

Reply:
I move the code in a function but I put the function inside Http namespace because the function is not related to authentication and it seems a bit weird for me to have such a "low level" function calling something in WebApplication class. Moreover, nearly everything I manipulate and overwrite are in Http namespace.

Reply:
I can't approve it. Env should be the "real" environment. Any substitutions should be done where it's needed.
Connection class can be used not only by WebApplication so this "corrupted" environment can be unexpected there.

Reply:
Connection class may be used elsewhere (I only found one call not related to webapplication in src/base/bittorrent/tracker.cpp) but in anycase, this is no corruption of environment, it is a fix to properly handle special cases when qbittorrent is installed behind reverse-proxy. This fix only occurs on incoming HTTP connection. In case of tracker.cpp, without the patch, clientAddress is always erroneous (see my ASCII drawing below to understand why).

If no reverse-proxy IP is defined, nothing changes. If one is defined and X-Forwarded-For is properly filled in HTTP header, qbittorrent will use the correct client IP instead of an erroneous one.

Reply:
At least the option should have more informative description, not only "Reverse proxy address". E.g. group box "Use forwarded client address" around it and some description inside it.

Reply:
Also I'm not sure you chose the correct plsce for this option since it's not for WebUI only.

Reply:
> Also I'm not sure you chose the correct plsce for this option since it's not for WebUI only.

It is only usable in WebUI because it is the only place accepting HTTP connection, that's why I place the new field here.

> At least the option should have more informative description, not only "Reverse proxy address". E.g. group box "Use forwarded client address" around it and some description inside it.

I will add a tooltip on the reverse proxy address field, thanks for the suggestion.

Reply:
It's also used by embedded Tracker.

Reply:
> It's also used by embedded Tracker.

Indeed. However, if a reverse proxy is used, this means qbittorrent cannot accept directly connections from the internet thus there is no reason to enable embedded Tracker thus no conflict can happen.

I don't know if qbittorrent can setup a SOCKS connection to accept external connection but in such a case, it is the SOCKS server public IP which should be use and AFAIK, this value should always filled manually.

Reply:
>Indeed. However, if a reverse proxy is used, this means qbittorrent cannot accept directly connections from the internet thus there is no reason to enable embedded Tracker thus no conflict can happen.

Where is the logic?
You want to accept WebUI from some network location via reverse proxy but you think "there is no reason to enable embedded Tracker" for users from the same network location. Why?

Reply:
I just assumed bittorrent tracker does not use HTTP protocol thus it cannot go through reverse proxy and has to use a SOCKS server to accept incoming client connection. Am I wrong ?

Reply:
BitTorrent Tracker is HTTP service. See official BitTorrent specifications for e.g.

Reply:
ok, Found it but if I read properly the specifications I found on wiki.theory.org, a tracker request may contain an optional IP:
> The parameter is only needed in the case where the IP address that the request came in on is not the IP address of the client. This happens if the client is communicating to the tracker through a proxy (or a transparent web proxy/cache.)

This means without my patch,  if tracker service is provided through a reverse-proxy and the client does not send its IP, the tracker will use reverse-proxy IP as client IP instead of the correct IP.

I agree to what Chocobo1 said yesterday, this kind of network topology should not be widely encountered but the patch is very small and if no reverse-proxy is defined only 3 more lines of code are processed on incoming connection. In the worst case, it changes nothing and in the best case, qbittorrent works properly where it should not before.

Don't tell me I am the only one to have such kind of settings in the world :)

Reply:
>The parameter is only needed in the case where the IP address that the request came in on is not the IP address of the client. This happens if the client is communicating to the tracker through a proxy (or a transparent web proxy/cache.)

Dubiously... at least, the client should know that the tracker is behind the proxy. IMO, it isn't useful in case of server-side proxy. Otherwise client should always add this param... but IIRC the reverse proxy works transparently for the clients, isn't it? And thus it doesn't require any changes on the client side.

Anyway, my POV is if you change environment globally then the corresponding option should be in more appropriate place. Otherwise you should do it inside WebApplication.

Reply:
> but IIRC the reverse proxy works transparently for the clients, isn't it?

exactly, that's why it adds X-Forwarded-For in HTTP header.

> my POV is if you change environment globally then the corresponding option should be in more appropriate place.

As it is globally, what about moving my code into base/net/proxyconfigurationmanager.cpp and the reverse proxy into connection tab in a new fieldset named "Reverse Proxy" ?

Reply:
>what about moving my code into base/net/proxyconfigurationmanager.cpp 

Why? Just because of word "Proxy" in its name? But ProxyConfigurationManager is for different things. It's for "outgoing" proxies.
>the reverse proxy into connection tab in a new fieldset named "Reverse Proxy" ?

Seems OK for me.

Reply:
> Why? Just because of word "Proxy" in its name? But ProxyConfigurationManager is for different things. It's for "outgoing" proxies.

Yes, I just want to avoid creation of a new file.

Reply:
* reverse proxy field moved into connection tab
* code moved into newly created class in src/base/net/reverseproxy.*

Reply:
>Yes, I just want to avoid creation of a new file.

New file is not necessary.